File: 백준/Silver/15663. N과 M （9）/N과 M （9）.js
const input = require("fs")
  .readFileSync("/dev/stdin")
  .toString()
  .trim()
  .split("\n");

const [N, M] = input[0].split(" ").map(Number);
const nums = input[1]
  .split(" ")
  .map(Number)
  .sort((a, b) => a - b);
const sequence = [];
const visit = Array.from({ length: N }, () => false);
const answer = new Set();

function dfs(depth) {
  if (depth === M) {
    answer.add(sequence.join(" "));
    return;
  }

  for (let i = 0; i < N; i++) {
    if (!visit[i]) {
      visit[i] = true;
      sequence.push(nums[i]);
      dfs(depth + 1);
      sequence.pop();
      visit[i] = false;
    }
  }
}

dfs(0);

console.log([...answer].join("\n"));

File: 백준/Silver/15663. N과 M （9）/N과 M （9）.py
N, M = map(int, input().split())
nums = list(map(int, input().split()))
stack = []
visited = [0] * (N+1)
answer = []

nums.sort()


def dfs(depth):
    if depth == M:
        if ' '.join(map(str, stack)) not in answer:
            answer.append(' '.join(map(str, stack)))
        return

    for idx, num in enumerate(nums):
        if not visited[idx]:
            visited[idx] = 1
            stack.append(num)
            dfs(depth+1)
            stack.pop()
            visited[idx] = 0


dfs(0)
print(*answer, sep='\n')
